packagedb/find_source_repo.py

```diff
@@ -0,0 +1,456 @@
+#
+# Copyright (c) nexB Inc. and others. All rights reserved.
+# purldb is a trademark of nexB Inc.
+# SPDX-License-Identifier: Apache-2.0
+# See http://www.apache.org/licenses/LICENSE-2.0 for the license text.
+# See https://github.com/nexB/purldb for support or download.
+# See https://aboutcode.org for more information about nexB OSS projects.
+#
+
+import enum
+import logging
+import subprocess
+from typing import Generator
+from typing import List
+from urllib.parse import urlparse
+
+import requests
+from packageurl import PackageURL
+from packageurl.contrib.purl2url import get_download_url
+from packageurl.contrib.purl2url import purl2url
+from scancode.api import get_urls as get_urls_from_location
+
+from minecode.model_utils import add_package_to_scan_queue
+from minecode.visitors.maven import get_merged_ancestor_package_from_maven_package
+from packagedb.models import Package
+from packagedb.models import PackageContentType
+from packagedb.models import PackageSet
+
+logger = logging.getLogger(__name__)
+
+
+class URLDataReturnType(enum.Enum):
+    """
+    Return type for get_urls_from_text
+    """
+
+    url = "url" # This the final URL after redirects
+    text = "text" # This is the text of the response
+
+
+non_reachable_urls = []
+CACHE = {
+    #    url: data
+}
+
+
+def get_urls_from_text(text):
+    """
+    Return the URLs found in a text
+    """
+    lines = text.splitlines()
+    # location can be a list of lines
+    for url in get_urls_from_location(location=lines)["urls"]:
+        yield url["url"]
+
+
+def get_data_from_response(response, data_type=URLDataReturnType.text):
+    """
+    Return the data from a response
+    """
+    if not response:
+        return
+    data_by_type = {
+        URLDataReturnType.url: response.url,
+        URLDataReturnType.text: response.text,
+    }
+    if data_type in data_by_type:
+        return data_by_type[data_type]
+    else:
+        raise ValueError(f"Invalid data_type: {data_type}")
+
+
+def get_data_from_url(
+    url,
+    data_type=URLDataReturnType.text,
+    timeout=None,
+):
+    """
+    Return the text from URL
```

Comment:
Update docstring to mention how the `data_type` argument is used and what is the expected things returned when you use `URLDataReturnType.text` vs `URLDataReturnType.url`